gba/tools/search/search_internet.py

```diff
@@ -0,0 +1,319 @@
+    "Title": 3,
+}
+
+
+class SearchInternetTool(Tool):
+    name: str = "search_internet"
+
+    def __init__(
+        self,
+        llm_model: Llama3Instruct,
+        rerank_model: CrossEncoder,
+        searxng_endpoint: str,
+        top_k_documents: int = 3,
+        top_k_nodes_per_document: int = 5,
+        top_k_snippets: int | None = None,
+        num_nodes_rerank_per_document: int = 100,
+        min_score: float = 0.01,
+        fetch_webpage_multiplier: int = 2,
+        fetch_webpage_timeout: float = 5.0,
+        max_concurrent_requests: int = 10,
+        extractor: ContentExtractor | None = None,
+    ):
+        """Search the internet for information.
+
+        :param llm_model: The LLM model to use for generating responses.
+        :param rerank_model: The reranker model to use for reranking document nodes.
+        :param searxng_endpoint: The endpoint of the searxng service.
+        :param top_k_documents: The number of top documents to consider for node extraction.
+        :param top_k_nodes_per_document: The number of top nodes to use for response generation.
+        :param top_k_snippets: The number of top snippets to use for response generation.
+        :param num_nodes_rerank_per_document: The number of nodes to consider for reranking per document.
+        :param min_score: The minimum score for reranking nodes.
+        :param fetch_webpage_multiplier: The multiplier for the number of webpages to fetch.
+        :param fetch_webpage_timeout: The timeout for fetching webpages.
+        :param max_concurrent_requests: The maximum number of concurrent requests.
+        :param extractor: The content extractor to use for extracting relevant information.
+        """
+        self._llm_model = llm_model
+        self._reranker = rerank_model
+        self._searxng_endpoint = searxng_endpoint.strip("/")
+        self._top_k_documents = top_k_documents
+        self._top_k_nodes_per_document = top_k_nodes_per_document
+        self._num_nodes_rerank_per_document = num_nodes_rerank_per_document
+        self._top_k_snippets = top_k_snippets
+        self._min_score = min_score
+        self._fetch_webpage_multiplier = fetch_webpage_multiplier
+        self._fetch_webpage_timeout = fetch_webpage_timeout
+        self._extractor = extractor
+        self._query_rewriter = QueryRewriter(model=llm_model)
+        self._query_pool = ThreadPoolExecutor(max_workers=max_concurrent_requests)
+        self._synthesise_query_pattern = re.compile(
+            r"^(search (for|to))|(search the internet (for|to))|(search wikipedia (for|to))\s", re.IGNORECASE
+        )
+
+    def run(
+        self,
+        request: str,
+        task: str,
+        scratchpad: Scratchpad,
+        **kwargs,
+    ) -> str:
+        """Useful for up-to-date information on the internet."""
+
+        search_query = self._query_rewriter.rewrite(task)
+        synthesise_response_query = self._synthesise_query_pattern.sub("", task.strip())
```

Comment:
I changed the prompts for both the `search_internet` and `search_wikipedia` tools to accept tasks instead of queries. The regex is now only used for the queries supplied to the `ContentExtractor`.

I left the `ContentExtrator` as-is for now, as the results degraded when supplying the original task description. This is left as a follow-up task.